buffer/ring_growing_test.go

```diff
@@ -17,47 +17,211 @@ limitations under the License.
+				t.Fatalf("expected Cap to be %d, got %d", expected, actual)
+			}
+
+			x := 10
+			for i := 0; i < x; i++ {
+				if e, a := i, g.readable; e != a {
+					t.Fatalf("expected equal, got %#v, %#v", e, a)
+				}
+				g.WriteOne(i)
+			}
+
+			if expected, actual := x, g.Len(); expected != actual {
+				t.Fatalf("expected Len to be %d, got %d", expected, actual)
+			}
+			if expected, actual := 16, g.Cap(); expected != actual {
+				t.Fatalf("expected Cap to be %d, got %d", expected, actual)
+			}
+
+			read := 0
+			for g.readable > 0 {
+				v, ok := g.ReadOne()
+				if !ok {
+					t.Fatal("expected true")
+				}
+				if read != v {
+					t.Fatalf("expected %#v==%#v", read, v)
+				}
+				read++
+			}
+			if x != read {
+				t.Fatalf("expected to have read %d items: %d", x, read)
+			}
+			if expected, actual := 0, g.Len(); expected != actual {
+				t.Fatalf("expected Len to be %d, got %d", expected, actual)
+			}
+			if expected, actual := 16, g.Cap(); expected != actual {
+				t.Fatalf("expected Cap to be %d, got %d", expected, actual)
+			}
+		})
 	}
-	if g.readable != 0 {
-		t.Fatalf("expected readable to be zero: %d", g.readable)
+
+}
+
+func TestGrowth(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		ring        *Ring[int]
+		initialSize int
+		normalSize  int
+	}{
+		"implicit-zero": {
+			ring: new(Ring[int]),
+		},
+		"explicit-zero": {
+			ring:        NewRing[int](RingOptions{InitialSize: 0, NormalSize: 0}),
+			initialSize: 0,
+			normalSize:  0,
+		},
+		"smaller-initial-size": {
+			ring:        NewRing[int](RingOptions{InitialSize: 1, NormalSize: 2}),
+			initialSize: 1,
+			normalSize:  2,
+		},
+		"smaller-normal-size": {
+			ring:        NewRing[int](RingOptions{InitialSize: 2, NormalSize: 1}),
+			initialSize: 2,
+			normalSize:  1,
+		},
 	}
-	if 16 != g.n {
-		t.Fatalf("expected N to be 16: %d", g.n)
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			initialSize := test.initialSize
+			normalSize := test.normalSize
+			g := test.ring
+
+			if expected, actual := 0, g.Len(); expected != actual {
+				t.Fatalf("expected Len to be %d, got %d", expected, actual)
+			}
+			if expected, actual := initialSize, g.Cap(); expected != actual {
+				t.Fatalf("expected Cap to be %d, got %d", expected, actual)
+			}
+
+			x := 10
+			for i := 0; i < x; i++ {
+				if e, a := i, g.growing.readable; e != a {
+					t.Fatalf("expected equal, got %#v, %#v", e, a)
+				}
+				g.WriteOne(i)
+			}
+
+			if expected, actual := x, g.Len(); expected != actual {
+				t.Fatalf("expected Len to be %d, got %d", expected, actual)
+			}
+			if expected, actual := 16, g.Cap(); expected != actual {
+				t.Fatalf("expected Cap to be %d, got %d", expected, actual)
+			}
+
+			read := 0
+			for g.growing.readable > 0 {
+				v, ok := g.ReadOne()
+				if !ok {
+					t.Fatal("expected true")
+				}
+				if read != v {
+					t.Fatalf("expected %#v==%#v", read, v)
+				}
+				read++
```

Comment:
note to self, we write the sequence of the index for loop, 1 2 3  ... so we check we read exactly those values